totem/circles/mobile_api/mobile_filters.py

```diff
@@ -0,0 +1,210 @@
+            Prefetch(
+                "events",
+                queryset=CircleEvent.objects.filter(start__gte=timezone.now())
+                .order_by("start")
+                .prefetch_related("attendees"),
+                to_attr="upcoming_events",
+            ),
+        )
+        .annotate(subscriber_count=Count("subscribed", distinct=True))
+    )
+
+
+def upcoming_recommended_spaces(user: User | None, categories: list[str] | None = None, author: str | None = None):
+    spaces = (
+        Circle.objects.filter(events__start__gte=timezone.now())
+        .distinct()
+        .select_related("author")
+        .prefetch_related(
+            "categories",
+            "subscribed",
+            Prefetch(
+                "events",
+                queryset=CircleEvent.objects.filter(start__gte=timezone.now())
+                .order_by("start")
+                .prefetch_related("attendees"),
+                to_attr="upcoming_events",
+            ),
+        )
+        .annotate(subscriber_count=Count("subscribed", distinct=True))
+    )
+    if not user or not user.is_staff:
+        spaces = spaces.filter(published=True)
+    if categories:
+        spaces = spaces.filter(Q(categories__slug__in=categories) | Q(categories__name__in=categories))
+    if author:
+        spaces = spaces.filter(author__slug=author)
+    return spaces
+
+
+def upcoming_recommended_events(user: User | None, categories: list[str] | None = None, author: str | None = None):
+    events = (
+        CircleEvent.objects.filter(start__gte=timezone.now(), cancelled=False, listed=True)
+        .select_related("circle")
+        .prefetch_related(
+            "circle__author",
+            "circle__categories",
+            "circle__subscribed",
+            Prefetch(
+                "circle__events",
+                queryset=CircleEvent.objects.filter(start__gte=timezone.now())
+                .order_by("start")
+                .prefetch_related("attendees"),
+                to_attr="upcoming_events",
+            ),
+        )
+        .annotate(
+            attendee_count=Count("attendees", distinct=True),
+            subscriber_count=Count("circle__subscribed", distinct=True),
+        )
+        .order_by("start")
+    )
+    if not user or not user.is_staff:
+        events = events.filter(circle__published=True)
+    # are there any seats?
+    events = events.filter(attendee_count__lt=F("seats"))
+    # filter category
+    if categories:
+        events = events.filter(Q(circle__categories__slug__in=categories) | Q(circle__categories__name__in=categories))
+    # filter author
+    if author:
+        events = events.filter(circle__author__slug=author)
+    return events
+
+
+def event_detail_schema(event: CircleEvent, user: User):
+    space: Circle = event.circle
+    start = event.start
+
+    if user.is_authenticated:
+        if hasattr(space, "_prefetched_objects_cache") and "subscribed" in space._prefetched_objects_cache:
```

Comment:
Dirty... Reaching into a private attribute (yay Python!), is a code smell, but hey if it works. Ideally this logic would get moved up the call tree so it can be done in a QuerySet with prefetch. This SO gives an idea: https://stackoverflow.com/questions/79509460/should-i-reach-into-the-django-prefetched-objects-cache-to-solve-an-n1-query. That being said, adding an attribute through a query, like the answer says to, can really mess with the type checker 😭

Reply:
Also, when messing with private methods or attributes it's really important to make sure tests are hitting these paths since this stuff isn't in the 'official' API and won't always follow semver. Could break on a point update! In this case it's unlikely since a lot of people use this, plus there's a fallback for it, but good practice!